Update to Google Analytics 4

Nateowami opened the pull request:
This change was copied verbatim from the Google Analytics interface.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/sillsdev/web-xforge/1839)
<!-- Reviewable:end -->

nigel-wells reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

I assume you've also made the required adjustments inside Google Tag Manager to pass the data through to GA4?

> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1839#-:-NVSGobD7fSfxFOjMmpQ:b55wrrf)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @Nateowami)



<!-- Sent from Reviewable.io -->

Nateowami commented:
@nigel-wells I have never understood what Google Tag Manager is, despite seeing it referenced everywhere in relation to Google Analytics. I'm pretty sure it's configured correctly. Can we look at it together and make sure before merging?

nigel-wells reviewed (dismissed):
I've talked this over with Mieke, she says the code you have will send the data directly to Google Analytics without the need to setup anything specific inside of Google Tag Manager. You could proceed with what you have in this PR and it will work. What you would lose is the ability to compare data going into both Google Analytics 4 and Google Universal Analytics which you could do if you were using Google Tag Manager containers. 

At some point it would be good to get a Google Tag Manager container setup as then other events and triggers could be setup for tracking specific things in the app. Did you want to go through the process of getting the base of that setup now as part of this change or do later? Mieke could certainly help with the setup if we can get permission on your account to manage it. 

Happy to chat things through.

> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1839#-:-NVWbDks2A9dhqe-ZMHA:b44q7qx)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @Nateowami)



<!-- Sent from Reviewable.io -->

Nateowami commented:
I somehow missed that we configure Google Analytics in a number of places. I've updated the other locations, and added logic to sanitize URLs before they're sent, as we've already been doing with Bugsnag.

nigel-wells reviewed (changes requested):
> Reviewed 11 of 11 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1839)* status: all files reviewed, 10 unresolved discussions (waiting on @Nateowami)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/index.html` line 4 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/1839#-NVlzIiU85mLb5dPUIkw:-NVlzIiU85mLb5dPUIkx:bnggq9n) ([raw file](https://github.com/sillsdev/web-xforge/blob/bd0b2b4ffda11a90a0a36659e1c9447b5ccceb2c/src/SIL.XForge.Scripture/ClientApp/src/index.html#L4)):*
> ```HTML
> <html lang="en">
>   <head>
>     <script async src="https://www.googletagmanager.com/gtag/js?id=G-SVKBDV7K3Q"></script>
> ```

You'll want to remove the tag from here - I understand it can be set later using the `config` command.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/index.html` line 10 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/1839#-NVm-Xqw9_PMRBl1GxBF:-NVm-Xqw9_PMRBl1GxBG:bymeloq) ([raw file](https://github.com/sillsdev/web-xforge/blob/bd0b2b4ffda11a90a0a36659e1c9447b5ccceb2c/src/SIL.XForge.Scripture/ClientApp/src/index.html#L10)):*
> ```HTML
>         dataLayer.push(arguments);
>       }
>       gtag("js", new Date());
> ```

Apparently this may also trigger a page view and should be reserved for being set when you're ready to trigger something along with the config.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/index.html` line 12 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/1839#-NVlssiQ0FDhkb12mHZS:-NVlssiQ0FDhkb12mHZT:b-mzsmoh) ([raw file](https://github.com/sillsdev/web-xforge/blob/bd0b2b4ffda11a90a0a36659e1c9447b5ccceb2c/src/SIL.XForge.Scripture/ClientApp/src/index.html#L12)):*
> ```HTML
>       gtag("js", new Date());
> 
>       gtag("config", "G-SVKBDV7K3Q");
> ```

You will want to remove this as this as the ID needs to come from the environment config plus I "think" it will trigger a page view. Essentially you want this specific `config` command to only be called once but I suspect, with the minimal tag manager configuration, it will also automatically trigger a page view.Bas
___
*[`src/SIL.XForge.Scripture/ClientApp/src/app/app.component.ts` line 150 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/1839#-NVltPSrE0fSHLA2_CCn:-NVltPSrE0fSHLA2_CCo:b6igx8e) ([raw file](https://github.com/sillsdev/web-xforge/blob/bd0b2b4ffda11a90a0a36659e1c9447b5ccceb2c/src/SIL.XForge.Scripture/ClientApp/src/app/app.component.ts#L150)):*
> ```XML
>       );
>       this.subscribe(navEndEvent$, e => {
>         if (this.isAppOnline) {
> ```

This check is already part of the new `AnalyticsService` so no need to have it here
___
*[`src/SIL.XForge.Scripture/ClientApp/src/xforge-common/analytics.service.ts` line 18 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/1839#-NVm0dHl7ETjtGhuoRMw:-NVm0dHl7ETjtGhuoRMx:b-ud5rdz) ([raw file](https://github.com/sillsdev/web-xforge/blob/bd0b2b4ffda11a90a0a36659e1c9447b5ccceb2c/src/SIL.XForge.Scripture/ClientApp/src/xforge-common/analytics.service.ts#L18)):*
> ```XML
> })
> export class AnalyticsService {
>   constructor(private readonly pwaService: PwaService) {}
> ```

Based on comments for `index.html` you may need to change part of this to ensure the config is set (and may the date?) but only on first logging. I'm not 100% (due to GTM setup) if setting that will in itself trigger a page view meaning you may not then need to trigger the alternative `page_view` event.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/xforge-common/analytics.service.ts` line 27 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/1839#-NVlvwJs2Y4PDDmzviSk:-NVlvwJs2Y4PDDmzviSl:bnkahd7) ([raw file](https://github.com/sillsdev/web-xforge/blob/bd0b2b4ffda11a90a0a36659e1c9447b5ccceb2c/src/SIL.XForge.Scripture/ClientApp/src/xforge-common/analytics.service.ts#L27)):*
> ```XML
>   logNavigation(url: string): void {
>     const sanitizedUrl = sanitizeUrl(url);
>     this.logEvent('page_view', { page_path: sanitizedUrl });
> ```

The command is `event` and the type is `page_view`. You can then set params as seen here: https://developers.google.com/tag-platform/gtagjs/reference/events#page_view
___
*[`src/SIL.XForge.Scripture/ClientApp/src/xforge-common/analytics.service.ts` line 30 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/1839#-NVlvISk4rCmxGhq2aOg:-NVlvISk4rCmxGhq2aOh:b-yt5sk3) ([raw file](https://github.com/sillsdev/web-xforge/blob/bd0b2b4ffda11a90a0a36659e1c9447b5ccceb2c/src/SIL.XForge.Scripture/ClientApp/src/xforge-common/analytics.service.ts#L30)):*
> ```XML
>   }
> 
>   private logEvent(eventName: string, eventParams: EventParams): void {
> ```

I'd stay away from `event` unless you're dealing specifically with a Google Analytics "event". Google's docs refers to the first param as a "command" followed up with the tag and additional parameters. We may want to specifically trigger a real "event" at some point in the future - so I'd prefer we reserved wording for that situation.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/xforge-common/analytics.service.ts` line 40 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/1839#-NVludCK8ruGew811ZI2:-NVludCK8ruGew811ZI3:bjzawvw) ([raw file](https://github.com/sillsdev/web-xforge/blob/bd0b2b4ffda11a90a0a36659e1c9447b5ccceb2c/src/SIL.XForge.Scripture/ClientApp/src/xforge-common/analytics.service.ts#L40)):*
> ```XML
> 
> // redact access token from the hash
> function redactAccessToken(url: string): string {
> ```

Perhaps we should put all these methods in its own utils file. Perhaps have a service we can call with one method and multiple ENUM options to pass with a string as to what needs to be redacted? This would save us having to remember all the different method names. It may not be obvious to look for bugsnag redact code in the analytics service.
___
*[`src/SIL.XForge.Scripture/ClientApp/src/xforge-common/analytics.service.spec.ts` line 10 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/1839#-NVltfC22ftpHLisRYHD:-NVltfC22ftpHLisRYHE:bastsvg) ([raw file](https://github.com/sillsdev/web-xforge/blob/bd0b2b4ffda11a90a0a36659e1c9447b5ccceb2c/src/SIL.XForge.Scripture/ClientApp/src/xforge-common/analytics.service.spec.ts#L10)):*
> ```XML
> 
>   it('should redact the join key from URL', () => {
>     const url = 'https://example.com/join/123';
> ```

Also run a test (or change to) for URLs that have the language code at the end i.e. `/join/123/en`
___
*[`src/SIL.XForge.Scripture/Pages/Shared/_Layout.cshtml` line 18 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/1839#-NVm1oIt0UZwDiu5R4Ru:-NVm1oIt0UZwDiu5R4Rv:brjtnam) ([raw file](https://github.com/sillsdev/web-xforge/blob/bd0b2b4ffda11a90a0a36659e1c9447b5ccceb2c/src/SIL.XForge.Scripture/Pages/Shared/_Layout.cshtml#L18)):*
> ```razor
>             gtag('js', new Date());
> 
>             gtag('config', 'G-SVKBDV7K3Q');
> ```

Can this come from an environment variable as well?


<!-- Sent from Reviewable.io -->

nigel-wells reviewed (commented):
Did you want to pick this one back up again @Nateowami ? It would be good to get it across as GA4 is well underway now and I'm not sure how much longer before UA is dropped.

> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1839#-:-NekFOvJ-F7don2-OuvB:b-17qoas)* status: all files reviewed, 10 unresolved discussions (waiting on @Nateowami)



<!-- Sent from Reviewable.io -->

Nateowami reviewed (commented):
That would be great. Thanks.

And coming back to Reviewable, I see I had a draft response that may still be relevant:
> I think I finally figured out where the gtag documentation is: https://developers.google.com/analytics/devguides/collection/ga4?hl=en

> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1839#-:-NVmHQp85DQN2vMA6Jml:b-g4lr5k)* status: all files reviewed, 10 unresolved discussions (waiting on @nigel-wells)

___
*[`src/SIL.XForge.Scripture/ClientApp/src/xforge-common/analytics.service.ts` line 30 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/1839#-NVlvISk4rCmxGhq2aOg:-NVmEnnx8H-KqaoMhgB_:b-6hbhyr) ([raw file](https://github.com/sillsdev/web-xforge/blob/bd0b2b4ffda11a90a0a36659e1c9447b5ccceb2c/src/SIL.XForge.Scripture/ClientApp/src/xforge-common/analytics.service.ts#L30)):*
<details><summary><i>Previously, nigel-wells (Nigel Wells) wrote…</i></summary><blockquote>

I'd stay away from `event` unless you're dealing specifically with a Google Analytics "event". Google's docs refers to the first param as a "command" followed up with the tag and additional parameters. We may want to specifically trigger a real "event" at some point in the future - so I'd prefer we reserved wording for that situation.
</blockquote></details>

Ah, didn't know that. So maybe use the "command" terminology?
___
*[`src/SIL.XForge.Scripture/Pages/Shared/_Layout.cshtml` line 18 at r2](https://reviewable.io/reviews/sillsdev/web-xforge/1839#-NVm1oIt0UZwDiu5R4Ru:-NVmEgz2Ek7ox4mLTVrB:b12qsz8) ([raw file](https://github.com/sillsdev/web-xforge/blob/bd0b2b4ffda11a90a0a36659e1c9447b5ccceb2c/src/SIL.XForge.Scripture/Pages/Shared/_Layout.cshtml#L18)):*
<details><summary><i>Previously, nigel-wells (Nigel Wells) wrote…</i></summary><blockquote>

Can this come from an environment variable as well?
</blockquote></details>

Probably, but it didn't look like it would be trivial.


<!-- Sent from Reviewable.io -->

nigel-wells reviewed (commented):
Would you be happy for me to run through making these updates as you're going to be out of action for a bit? You can then review. I'll just stick to what we have here rather than introducing any other special events or tracking as that can be added later on.

> *[Reviewable](https://reviewable.io/reviews/sillsdev/web-xforge/1839#-:-NepRHzY4fWVZZ_hUIPW:b-p39y2n)* status: all files reviewed, 10 unresolved discussions (waiting on @Nateowami)



<!-- Sent from Reviewable.io -->

Nateowami commented:
Yes, that would be great.

nigel-wells commented:
I have this mostly working now with data flowing through to GA4. It is only triggering, however, on first page load. I suspect this is a setting that needs to be tweaked in the Google Tag Manager instance to work better with SPA instances. Am I able to get permission to view the settings in there as well please?

nigel-wells commented:
I've taken another look at this and I see I can access the Google Tag Manager screen for the account which I didn't think I had access to. I'll take another look at this setup and will come back to you with next steps.

nigel-wells commented:
@Nateowami I've pushed through where I was at with this. I know, at the time, this was all working. 

I've rebased it off `master` but not done any testing other than making sure it builds. There will likely be fixes needed to be made as a number of changes have been made to the repo since this work began.

The hold up was needing admin access to Google Analytics so that it could be linked up to some new Google Tag Manager containers that I setup. 

It would still be good for someone to get this across the line as there is a lot of opportunity and insights to be gained by having better data in Google Analytics.
